File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/reader/ChapterReader.java
package com.github.doomsdayrs.apps.shosetsu.ui.reader;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.github.doomsdayrs.api.shosetsu.services.core.dep.Formatter;
import com.github.doomsdayrs.apps.shosetsu.R;
import com.github.doomsdayrs.apps.shosetsu.backend.Utilities;
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database;
import com.github.doomsdayrs.apps.shosetsu.ui.reader.adapters.ChapterReaderAdapter;
import com.github.doomsdayrs.apps.shosetsu.ui.reader.demarkActions.IndentChange;
import com.github.doomsdayrs.apps.shosetsu.ui.reader.demarkActions.ParaSpacingChange;
import com.github.doomsdayrs.apps.shosetsu.ui.reader.demarkActions.ReaderChange;
import com.github.doomsdayrs.apps.shosetsu.ui.reader.demarkActions.TextSizeChange;
import com.github.doomsdayrs.apps.shosetsu.variables.DefaultScrapers;
import com.github.doomsdayrs.apps.shosetsu.variables.Settings;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.isReaderNightMode;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.isTapToScroll;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.openInBrowser;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.openInWebview;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.setTextSize;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.setupTheme;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.swapReaderColor;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.toggleBookmarkChapter;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.toggleTapToScroll;
import static com.github.doomsdayrs.apps.shosetsu.backend.Utilities.unmarkMenuItems;
import static com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.getChaptersOnlyIDs;
import static com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseNovels.getReaderType;
import static java.util.Arrays.binarySearch;

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 */

/**
 * Shosetsu
 * 9 / June / 2019
 *
 * @author github.com/doomsdayrs
 */
//TODO MarkDown support
public class ChapterReader extends AppCompatActivity {
    private final Utilities.DemarkAction[] demarkActions = {new TextSizeChange(null), new ParaSpacingChange(null), new IndentChange(null), new ReaderChange(null)};
    public int[] chapterIDs;
    public int currentChapterID;
    public boolean first = false;
    public final ArrayList<ChapterView> chapters = new ArrayList<>();


    // Order of values. Small,Medium,Large
    private final MenuItem[] textSizes = new MenuItem[3];
    // Order of values. Non,Small,Medium,Large
    private final MenuItem[] paragraphSpaces = new MenuItem[4];
    // Order of values. Non,Small,Medium,Large
    private final MenuItem[] indentSpaces = new MenuItem[4];
    // Order of values. Default, Markdown
    private final MenuItem[] readers = new MenuItem[2];

    public Toolbar toolbar;


    private ViewPager chapterPager;

    public Formatter formatter;
    private MenuItem bookmark;
    public int readerType;
    private int novelID;
    public ChapterView currentView;

    private MenuItem tap_to_scroll;

    /**
     * Save data of view before destroyed
     *
     * @param outState output save
     */
    @Override
    protected void onSaveInstanceState(@NotNull Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putInt("formatter", formatter.getFormatterID());
    }

    /**
     * Creates the option menu (on the top toolbar)
     *
     * @param menu Menu reference to fill
     * @return if made
     */
    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        MenuInflater inflater = getMenuInflater();

        inflater.inflate(R.menu.toolbar_chapter_view, menu);
        // Night mode
        menu.findItem(R.id.chapter_view_nightMode).setChecked(isReaderNightMode());

        // Bookmark
        {
            bookmark = menu.findItem(R.id.chapter_view_bookmark);
            currentView.bookmarked = Database.DatabaseChapter.isBookMarked(currentView.chapterID);
            updateBookmark();
        }

        // Tap To Scroll
        {
            tap_to_scroll = menu.findItem(R.id.tap_to_scroll);
            tap_to_scroll.setChecked(isTapToScroll());
        }
        // Text size
        {
            textSizes[0] = menu.findItem(R.id.chapter_view_textSize_small);
            textSizes[1] = menu.findItem(R.id.chapter_view_textSize_medium);
            textSizes[2] = menu.findItem(R.id.chapter_view_textSize_large);

            switch ((int) Settings.ReaderTextSize) {
                default:
                    setTextSize(14);
                case 14:
                    textSizes[0].setChecked(true);
                    break;
                case 17:
                    textSizes[1].setChecked(true);
                    break;
                case 20:
                    textSizes[2].setChecked(true);
                    break;
            }
        }

        // Paragraph Space
        {
            paragraphSpaces[0] = menu.findItem(R.id.chapter_view_paragraphSpace_none);
            paragraphSpaces[1] = menu.findItem(R.id.chapter_view_paragraphSpace_small);
            paragraphSpaces[2] = menu.findItem(R.id.chapter_view_paragraphSpace_medium);
            paragraphSpaces[3] = menu.findItem(R.id.chapter_view_paragraphSpace_large);

            paragraphSpaces[Settings.paragraphSpacing].setChecked(true);
        }

        // Indent Space
        {
            indentSpaces[0] = menu.findItem(R.id.chapter_view_indent_none);
            indentSpaces[1] = menu.findItem(R.id.chapter_view_indent_small);
            indentSpaces[2] = menu.findItem(R.id.chapter_view_indent_medium);
            indentSpaces[3] = menu.findItem(R.id.chapter_view_indent_large);

            indentSpaces[Settings.indentSize].setChecked(true);
        }

        // Reader
        {

            readers[0] = menu.findItem(R.id.reader_0);
            readers[1] = menu.findItem(R.id.reader_1);
            readerType = getReaderType(novelID);

            switch (readerType) {
                case 1:
                    unmarkMenuItems(readers, 1, null);
                    break;
                case 0:
                case -1:
                    unmarkMenuItems(readers, 0, null);
                    break;
                case -2:
                default:
                    throw new RuntimeException("Invalid chapter?!? How are you reading this without the novel loaded in");
            }
        }
        return true;
    }

    public void updateBookmark() {
        if (bookmark != null)
            if (currentView.bookmarked)
                bookmark.setIcon(R.drawable.ic_bookmark_black_24dp);
            else bookmark.setIcon(R.drawable.ic_bookmark_border_black_24dp);
    }

    /**
     * What to do when an menu item is selected
     *
     * @param item item selected
     * @return true if processed
     */
    @Override
    public boolean onOptionsItemSelected(@NotNull MenuItem item) {
        Log.d("item", item.toString());
        switch (item.getItemId()) {
            default:
                return false;
            case R.id.chapter_view_nightMode:
                if (!item.isChecked()) {
                    swapReaderColor();
                    currentView.setUpReader();
                } else {
                    swapReaderColor();
                    currentView.setUpReader();
                }
                item.setChecked(!item.isChecked());
                return true;
            case R.id.tap_to_scroll:
                tap_to_scroll.setChecked(toggleTapToScroll());
                return true;
            case R.id.chapter_view_bookmark:
                currentView.bookmarked = toggleBookmarkChapter(currentView.chapterID);
                updateBookmark();
                return true;
            case R.id.chapter_view_textSize_small:
                unmarkMenuItems(indentSpaces, 0, demarkActions[0]);
                return true;
            case R.id.chapter_view_textSize_medium:
                unmarkMenuItems(textSizes, 1, demarkActions[0]);
                return true;
            case R.id.chapter_view_textSize_large:
                unmarkMenuItems(textSizes, 2, demarkActions[0]);
                return true;

            case R.id.chapter_view_paragraphSpace_none:
                unmarkMenuItems(paragraphSpaces, 0, demarkActions[1]);
                return true;
            case R.id.chapter_view_paragraphSpace_small:
                unmarkMenuItems(paragraphSpaces, 1, demarkActions[1]);
                return true;
            case R.id.chapter_view_paragraphSpace_medium:
                unmarkMenuItems(paragraphSpaces, 2, demarkActions[1]);
                return true;
            case R.id.chapter_view_paragraphSpace_large:
                unmarkMenuItems(paragraphSpaces, 3, demarkActions[1]);
                return true;

            case R.id.chapter_view_indent_none:
                unmarkMenuItems(indentSpaces, 0, demarkActions[2]);
                return true;
            case R.id.chapter_view_indent_small:
                unmarkMenuItems(indentSpaces, 1, demarkActions[2]);
                return true;
            case R.id.chapter_view_indent_medium:
                unmarkMenuItems(indentSpaces, 2, demarkActions[2]);
                return true;
            case R.id.chapter_view_indent_large:
                unmarkMenuItems(indentSpaces, 3, demarkActions[2]);
                return true;

            case R.id.browser:
                openInBrowser(this, currentView.chapterURL);
                return true;
            case R.id.webview:
                openInWebview(this, currentView.chapterURL);
                return true;
            case R.id.reader_0:
                unmarkMenuItems(readers, 0, demarkActions[3]);
                return true;
            case R.id.reader_1:
                unmarkMenuItems(readers, 1, demarkActions[3]);
                return true;

        }
    }

    /**
     * Create method
     *
     * @param savedInstanceState save object
     */
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d("OnCreate", "NovelFragmentChapterReader");
        setupTheme(this);
        setContentView(R.layout.chapter_reader);
        // SetUp of data

        if (savedInstanceState != null) {
            formatter = DefaultScrapers.getByID(savedInstanceState.getInt("formatter"));
            novelID = savedInstanceState.getInt("novelID");
            chapterIDs = savedInstanceState.getIntArray("chapters");
        } else {
            chapterIDs = getIntent().getIntArrayExtra("chapters");
            {
                int chapterID;
                chapterID = getIntent().getIntExtra("chapterID", -1);
                currentChapterID = chapterID;
            }
            novelID = getIntent().getIntExtra("novelID", -1);
            formatter = DefaultScrapers.getByID(getIntent().getIntExtra("formatter", -1));
        }


        if (chapterIDs == null) {
            List<Integer> integers = getChaptersOnlyIDs(novelID);
            chapterIDs = new int[integers.size()];
            int y = 0;
            for (int x = 1; x < integers.size(); x++) {
                chapterIDs[x] = integers.get(y);
                y++;
            }
        }

        // Declares view variables
        {
            toolbar = findViewById(R.id.toolbar);
            chapterPager = findViewById(R.id.viewpager);
        }
        setSupportActionBar(toolbar);
        setViewPager();
    }

    private void setViewPager() {
        ChapterReaderAdapter pagerAdapter = new ChapterReaderAdapter(getSupportFragmentManager(), FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT, this);
        chapterPager.setAdapter(pagerAdapter);
        chapterPager.setCurrentItem(getPosition());
    }

    private int getPosition() {
        return binarySearch(chapterIDs, currentChapterID);
    }


    public ChapterView cachedChapter(int chapterID) {
        for (ChapterView chapterView : chapters)
            if (chapterView.chapterID == chapterID)
                return chapterView;
        return null;
    }
}

File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/reader/ChapterReader.kt
package com.github.doomsdayrs.apps.shosetsu.ui.reader

import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.util.set
import androidx.core.view.postDelayed
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DiffUtil.Callback
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.github.doomsdayrs.apps.shosetsu.R
import com.github.doomsdayrs.apps.shosetsu.common.Settings
import com.github.doomsdayrs.apps.shosetsu.common.Settings.MarkingTypes
import com.github.doomsdayrs.apps.shosetsu.common.consts.BundleKeys.BUNDLE_CHAPTER_ID
import com.github.doomsdayrs.apps.shosetsu.common.consts.BundleKeys.BUNDLE_NOVEL_ID
import com.github.doomsdayrs.apps.shosetsu.common.dto.HResult
import com.github.doomsdayrs.apps.shosetsu.common.enums.ReadingStatus.READING
import com.github.doomsdayrs.apps.shosetsu.common.ext.logID
import com.github.doomsdayrs.apps.shosetsu.common.ext.observe
import com.github.doomsdayrs.apps.shosetsu.common.ext.openInBrowser
import com.github.doomsdayrs.apps.shosetsu.common.ext.openInWebView
import com.github.doomsdayrs.apps.shosetsu.ui.reader.adapters.ChapterReaderAdapter
import com.github.doomsdayrs.apps.shosetsu.view.uimodels.ReaderChapterUI
import com.github.doomsdayrs.apps.shosetsu.viewmodel.base.IChapterReaderViewModel
import com.sothree.slidinguppanel.SlidingUpPanelLayout
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState
import com.xw.repo.BubbleSeekBar
import kotlinx.android.synthetic.main.chapter_reader.*
import kotlinx.android.synthetic.main.chapter_reader_bottom.*
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.android.closestKodein
import org.kodein.di.generic.instance

/*
 * This file is part of shosetsu.
 *
 * shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 */ /**
 * shosetsu
 * 13 / 12 / 2019
 *
 * @author github.com/doomsdayrs
 */
class ChapterReader
	: AppCompatActivity(R.layout.chapter_reader), KodeinAware {
	private class RecyclerViewDiffer(
			val old: List<ReaderChapterUI>,
			val aNew: List<ReaderChapterUI>
	) : Callback() {
		override fun getOldListSize(): Int = old.size

		override fun getNewListSize(): Int = aNew.size

		override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
				old[oldItemPosition].id == aNew[newItemPosition].id

		override fun areContentsTheSame(
				oldItemPosition: Int,
				newItemPosition: Int
		): Boolean {
			val o: ReaderChapterUI = old[oldItemPosition]
			val n = aNew[newItemPosition]
			if (o.id != n.id) return false
			if (o.link != n.link) return false
			if (o.title != n.title) return false
			return true
		}
	}

	// Order of values. Night, Light, Sepia
	private lateinit var themes: Array<MenuItem>

	override val kodein: Kodein by closestKodein()
	internal val viewModel by instance<IChapterReaderViewModel>()
	private val chapterReaderAdapter: ChapterReaderAdapter = ChapterReaderAdapter(this)
	private val pageChangeCallback: OnPageChangeCallback = object : OnPageChangeCallback() {
		override fun onPageSelected(position: Int) {
			Log.d(logID(), "Page changed to $position ${chapters[position].link}")
			val chapterReaderUI = chapters[position]
			viewModel.currentChapterID = chapterReaderUI.id

			// Mark read if set to onview
			if (Settings.readerMarkingType == MarkingTypes.ONVIEW) {
				Log.d("ChapterReader", "Marking as reading by marking type")
				chapterReaderUI.readingStatus = READING
				viewModel.updateChapter(chapterReaderUI)
			}

			supportActionBar?.title = chapterReaderUI.title

			bookmark.setImageResource(
					if (chapterReaderUI.bookmarked)
						R.drawable.ic_bookmark_24dp
					else R.drawable.ic_bookmark_border_24dp
			)
		}
	}

	/**
	 * Chapters to display owo
	 */
	val chapters: ArrayList<ReaderChapterUI> = arrayListOf()

	public override fun onCreate(savedInstanceState: Bundle?) {
		Log.d(logID(), "On Create")
		window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_IMMERSIVE
		super.onCreate(savedInstanceState)
		setSupportActionBar(toolbar as Toolbar)
		supportActionBar?.setDisplayHomeAsUpEnabled(true)
		slidingUpPanelLayout.setGravity(Gravity.BOTTOM)

		viewModel.apply {
			setNovelID(intent.getIntExtra(BUNDLE_NOVEL_ID, -1))
			currentChapterID = intent.getIntExtra(BUNDLE_CHAPTER_ID, -1)
		}

		setObservers()
		setupViewPager()
		setupBottomMenu()
	}

	override fun onDestroy() {
		Log.d(logID(), "Destroying")
		super.onDestroy()
		viewpager.unregisterOnPageChangeCallback(pageChangeCallback)
	}

	private fun setObservers() {
		viewModel.liveData.observe(this) {
			when (it) {
				is HResult.Loading -> {
					Log.d(logID(), "Loading")
				}
				is HResult.Empty -> {
				}
				is HResult.Error -> {
				}
				is HResult.Success -> {
					Log.d(logID(), "Loading complete, now displaying")

					val currentSize = chapters.size
					val dif = DiffUtil.calculateDiff(RecyclerViewDiffer(
							chapters,
							it.data
					))
					chapters.clear()
					chapters.addAll(it.data)
					if (currentSize == 0) {
						getCurrentChapter()?.let {
							supportActionBar?.title = it.title
						}
						setupViewPager()
					}
					dif.dispatchUpdatesTo(chapterReaderAdapter)
					//bookmark.setIcon(if (it.data.bookmarked)
					//	R.drawable.ic_bookmark_24dp
					//else
					//	R.drawable.ic_bookmark_border_24dp
					//)
				}
			}
		}
	}

	private fun getCurrentChapter() = chapters.find {
		it.id == viewModel.currentChapterID
	}

	private fun getCurrentChapterIndex() = chapters.indexOfFirst {
		it.id == viewModel.currentChapterID
	}

	private fun setBookmarkIcon(readerChapterUI: ReaderChapterUI) {
		bookmark.setImageResource(
				if (readerChapterUI.bookmarked)
					R.drawable.ic_bookmark_24dp
				else R.drawable.ic_bookmark_border_24dp
		)
	}

	private fun setupBottomMenu() {
		bookmark.apply {
			setOnClickListener {
				getCurrentChapter()?.apply {
					bookmarked = !bookmarked
					viewModel.updateChapter(this)
					setBookmarkIcon(this)
				}
			}
		}

		text_size_bar?.apply {
			setCustomSectionTextArray { _, array ->
				array.apply {
					clear()
					this[0] = getString(R.string.small)
					this[1] = getString(R.string.medium)
					this[2] = getString(R.string.large)
				}
			}
			onProgressChangedListener = object : BubbleSeekBar.OnProgressChangedListener {
				override fun onProgressChanged(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float,
						fromUser: Boolean
				) {
					if (fromUser) {
						val size = when (progress) {
							0 -> Settings.TextSizes.SMALL
							1 -> Settings.TextSizes.MEDIUM
							2 -> Settings.TextSizes.LARGE
							else -> Settings.TextSizes.MEDIUM
						}
						Log.i(logID(), "TextSize changed to ${size.name}")
						Settings.readerTextSize = size.i
						// Sets current view
						chapterReaderAdapter.textReaders.find {
							it.chapterID == viewModel.currentChapterID
						}?.let {
							it.textView.textSize = size.i
						}
						// Sets other views down
						chapterReaderAdapter.textReaders.filter {
							it.chapterID != viewModel.currentChapterID
						}.forEach {
							it.textView.textSize = size.i
						}
					}
				}

				override fun getProgressOnActionUp(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float
				) {
				}

				override fun getProgressOnFinally(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float,
						fromUser: Boolean
				) {

				}

			}
		}

		para_space_bar?.apply {
			setCustomSectionTextArray { _, array ->
				array.apply {
					clear()
					this[0] = getString(R.string.none)
					this[1] = getString(R.string.small)
					this[2] = getString(R.string.medium)
					this[3] = getString(R.string.large)
				}
			}
			onProgressChangedListener = object : BubbleSeekBar.OnProgressChangedListener {
				override fun onProgressChanged(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float,
						fromUser: Boolean
				) {
					if (fromUser) {
						Log.i(logID(), "ParaSpace changed to $progress")
						Settings.readerParagraphSpacing = progress
						// Sets current view
						chapterReaderAdapter.textReaders.find {
							it.chapterID == viewModel.currentChapterID
						}?.let {
							it.bind()
						}
						// Sets other views down
						chapterReaderAdapter.textReaders.filter {
							it.chapterID != viewModel.currentChapterID
						}.forEach {
							it.bind()
						}
					}
				}

				override fun getProgressOnActionUp(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float
				) {
				}

				override fun getProgressOnFinally(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float,
						fromUser: Boolean
				) {

				}

			}
		}

		para_indent_bar?.apply {
			setCustomSectionTextArray { _, array ->
				array.apply {
					clear()
					this[0] = getString(R.string.none)
					this[1] = getString(R.string.small)
					this[2] = getString(R.string.medium)
					this[3] = getString(R.string.large)
				}
			}
			onProgressChangedListener = object : BubbleSeekBar.OnProgressChangedListener {
				override fun onProgressChanged(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float,
						fromUser: Boolean
				) {
					if (fromUser) {
						Log.i(logID(), "IndentSize changed to $progress")
						Settings.readerIndentSize = progress
						// Sets current view
						chapterReaderAdapter.textReaders.find {
							it.chapterID == viewModel.currentChapterID
						}?.let {
							it.bind()
						}
						// Sets other views down
						chapterReaderAdapter.textReaders.filter {
							it.chapterID != viewModel.currentChapterID
						}.forEach {
							it.bind()
						}
					}
				}

				override fun getProgressOnActionUp(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float
				) {
				}

				override fun getProgressOnFinally(
						bubbleSeekBar: BubbleSeekBar?,
						progress: Int,
						progressFloat: Float,
						fromUser: Boolean
				) {

				}

			}
		}
	}

	private fun setupViewPager() {
		Log.d(logID(), "Setting up ViewPager")
		viewpager.apply {
			adapter = chapterReaderAdapter
			registerOnPageChangeCallback(pageChangeCallback)
			currentItem = getCurrentChapterIndex()
			addItemDecoration(DividerItemDecoration(
					viewpager.context,
					viewpager.orientation
			))
		}
	}

	fun animateBottom() {
		chapter_reader_bottom?.apply {
			post {
				slidingUpPanelLayout.apply {
					val currentState = panelState!!
					Log.d(logID(), "Changing panelState from ${currentState.name} | $panelHeight")
					val state = when (currentState) {
						PanelState.HIDDEN -> {
							Log.d(logID(), "Hidden, making collapsed")
							PanelState.COLLAPSED
						}
						PanelState.ANCHORED -> {
							Log.d(logID(), "ANCHORED, making collapsed")
							PanelState.COLLAPSED
						}
						PanelState.COLLAPSED -> {
							Log.d(logID(), "COLLAPSED, making hidden")
							PanelState.HIDDEN
						}
						PanelState.DRAGGING -> {
							Log.d(logID(), "Dragging, making hidden")
							PanelState.HIDDEN
						}
						PanelState.EXPANDED -> {
							Log.d(logID(), "Expanded, making hidden")
							PanelState.HIDDEN
						}
					}
					panelState = state
					postDelayed(400) { fixHeight() }
				}
			}
		}
	}

	private fun fixHeight(): SlidingUpPanelLayout = slidingUpPanelLayout.apply {
		val state = panelState
		Log.d(logID(), "PanelState is now ${state.name}")
		when (state) {
			PanelState.HIDDEN -> panelHeight = 0
			PanelState.COLLAPSED -> panelHeight = 238
			PanelState.DRAGGING -> postDelayed(100) { fixHeight() }
			else -> Log.d(logID(), "Unknown state: $state")
		}
	}

	fun animateToolbar() {
		toolbar?.let {
			@Suppress("CheckedExceptionsKotlin")
			val animator: Animation = AnimationUtils.loadAnimation(
					it.context,
					if (it.visibility == VISIBLE)
						R.anim.slide_up
					else R.anim.slide_down
			)
			it.startAnimation(animator)
			it.post {
				it.visibility = if (it.visibility == VISIBLE) GONE else VISIBLE
			}
		}

	}

	override fun onCreateOptionsMenu(menu: Menu): Boolean {
		menuInflater.inflate(R.menu.toolbar_chapter_view, menu)
		// Night mode
		run {
			themes = arrayOf(
					menu.findItem(R.id.chapter_view_reader_night),
					menu.findItem(R.id.chapter_view_reader_light),
					menu.findItem(R.id.chapter_view_reader_sepia),
					menu.findItem(R.id.chapter_view_reader_dark),
					menu.findItem(R.id.chapter_view_reader_gray),
					menu.findItem(R.id.chapter_view_reader_custom)

			)
			when (Settings.readerTheme) {
				0 -> themes[0].setChecked(true)
				1 -> themes[1].setChecked(true)
				2 -> themes[2].setChecked(true)
				3 -> themes[3].setChecked(true)
				4 -> themes[4].setChecked(true)
				5 -> themes[5].setChecked(true)
				else -> {
					Settings.readerTheme = 1
					themes[1].setChecked(true)
				}
			}
		}
		return true
	}

	/**
	 * What to do when an menu item is selected
	 *
	 * @param item item selected
	 * @return true if processed
	 */
	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		Log.d(logID(), "Selected item: $item")
		return when (item.itemId) {
			android.R.id.home -> {
				finish()
				true
			}
			R.id.chapter_view_reader_night -> {
				true
			}
			R.id.chapter_view_reader_light -> {
				true
			}
			R.id.chapter_view_reader_sepia -> {
				true
			}
			R.id.chapter_view_reader_dark -> {
				true
			}
			R.id.chapter_view_reader_gray -> {
				true
			}
			R.id.chapter_view_reader_custom -> {
				true
			}
			R.id.browser -> {
				val url = chapters[viewModel.currentChapterID].link
				if (url.isNotEmpty())
					openInBrowser(url)
				true
			}
			R.id.webview -> {
				val url = chapters[viewModel.currentChapterID].link
				if (url.isNotEmpty())
					openInWebView(url)
				true
			}
			else -> super.onOptionsItemSelected(item)
		}
	}
}